homeassistant/components/dyson/air_quality.py

```diff
@@ -0,0 +1,146 @@
+
+    def __init__(self, device):
+        """Create a new generic air quality Dyson sensor."""
+        self._device = device
+        self._old_value = None
+        self._name = device.name
+
+    async def async_added_to_hass(self):
+        """Call when entity is added to hass."""
+        self.hass.async_add_executor_job(
+            self._device.add_message_listener, self.on_message)
+
+    def on_message(self, message):
+        """Handle new messages which are received from the fan."""
+        # Prevent refreshing if not needed
+        if self._old_value is None or self._old_value != self.state:
+            _LOGGER.debug("Message received for %s device: %s", self.name,
+                          message)
+            self._old_value = self.state
+            self.schedule_update_ha_state()
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the Dyson sensor."""
+        return self._name
+
+    @property
+    def attribution(self):
+        """Return the attribution."""
+        return ATTRIBUTION
+
+    @property
+    def air_quality_index(self):
+        """Return the Air Quality Index (AQI)."""
+        return max(self.particulate_matter_2_5,
+                   self.particulate_matter_10,
+                   self.nitrogen_dioxide,
+                   self.volatile_organic_compounds)
+
+    @property
+    def particulate_matter_2_5(self):
+        """Return the particulate matter 2.5 level."""
+        if self._device.environmental_state:
+            return int(self._device.environmental_state.particulate_matter_25)
+
+    @property
+    def particulate_matter_10(self):
+        """Return the particulate matter 10 level."""
+        if self._device.environmental_state:
+            return int(self._device.environmental_state.particulate_matter_10)
+
+    @property
+    def nitrogen_dioxide(self):
+        """Return the NO2 (nitrogen dioxide) level."""
+        if self._device.environmental_state:
+            return int(self._device.environmental_state.nitrogen_dioxide)
+
+    @property
+    def volatile_organic_compounds(self):
+        """Return the VOC (Volatile Organic Compounds) level."""
+        if self._device.environmental_state:
+            return int(self._device.
+                       environmental_state.volatile_organic_compounds)
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        data = {}
+
+        for prop, attr in PROP_TO_ATTR.items():
+            value = getattr(self, prop)
+            if value is not None:
+                data[attr] = value
+
+        return data
+
+    @property
+    def state(self):
+        """Return the current state."""
+        return self.air_quality_index
+
+    @property
+    def device_serial(self):
+        """Return the parent fan's serial number."""
+        return self._device.serial
```

Comment:
Is the serial unique? Then we could use it for `unique_id` property.

https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements

Reply:
Yep, the serial should be unique for every device, thanks for the tip, I'll use the unique_id property instead of adding another 1 for the serial.